### PR TITLE
Xml parser

### PR DESCRIPTION
**Pull request by LaurensVanRijssel**

### CODE REVIEW

File: code/sourceReceiver.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
+# Generate sources and receivers
+import math
+import xml.etree.cElementTree as ET
+import matplotlib.pyplot as plt
+import numpy as np
```
Comment (Constantijn-Dinklo) [inline code/sourceReceiver.py]: Not a file that belongs to this commit.

File: code/misc.py [hunk 2]
```diff
@@ -19,9 +19,9 @@ def side_test(pa, pb, pc):
         the computed value is actually the 2* area of the triangle
         ---------------
         Input:
-        pa = [x,y,z]
-        pb = [x,y,z]
-        pc = [x,y,z]
+        pa = [x,y(,z)]
+        pb = [x,y(,z)]
+        pc = [x,y(,z)]
```
Comment (Constantijn-Dinklo) [inline code/misc.py:24]: Why is a Z value included at all? It is never used.

File: code/groundTin.py [hunk 3]
```diff
@@ -214,8 +214,6 @@ def create_cross_section(self, edge_list, source, receiver, source_tr, receiver_
         receiver_height = self.interpolate_triangle(receiver_tr, receiver)
         cross_section_vertices = [(receiver[0], receiver[1], receiver_height)]
 
-        # the edges go from 0 - 1, 1-2, 2-3, 3-4, 4-5, 5-6
-        cross_section_edges = [[0]]
         for i, edge in enumerate(edge_list):
             # get the vertices (from the vts list) and turn them into a numpt array for vector processing.
             vertex_right = np.array(self.vts[edge[0]])
```
Comment (Constantijn-Dinklo) [inline code/groundTin.py]: Do these file changes belong to this commit? They seem irrelevant. We don't want to take part of a change, or something you were testing.

File: code/xmlParser.py [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+
+    def set_coordinates_local(self):
+        """
+        Explination: Move 3D Casrtesian coordinates relative to the starting point (receiver) by subtraction P0 from everypoint
+        ---------------
+        Input: void
+        ---------------
+        Output: void
+        """
+        self.vts -= self.vts[0]
```
Comment (Constantijn-Dinklo) [inline code/xmlParser.py]: This is generally called 'normalize'. But just a suggestion.

File: code/xmlParser.py [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+
+    def unfold_straight_path(self):
+        """
+        Explination: calculate the pythagoras distance and make a new array with the distance d and the height z
+        ---------------
+        Input: void
+        ---------------
+        Output: void (fills self.vts_dz)
```
Comment (Constantijn-Dinklo) [inline code/xmlParser.py]: self.vts_dz is never accessed in this method and can thus not be filled.

File: code/xmlParser.py [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+        """
+        # calc the diagonal distance for all vertices
+        d = (self.vts[:,0] ** 2 + self.vts[:,1] ** 2) ** 0.5
+
+        y = np.zeros(len(self.vts), dtype=float)
+
+        # stack the diagonal distance with the height
+        self.vts = np.vstack((d, y, self.vts[:,2])).T
```
Comment (Constantijn-Dinklo) [inline code/xmlParser.py]: Are you swapping the values in self.vts from [x,y,z] to [d, y, z]? What is the relevance of the y value here?
Might need to go into the self.vts_dz list?

File: code/xmlParser.py [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+        """
+        Explination:
+            calculates the perpendicular distance from points to the the line
+        ---------------
+        Input: 
+            Start = id of the start point
+            end = id of the end point
+        ---------------
+        Output: 
+            List of distances
```
Comment (Constantijn-Dinklo) [inline code/xmlParser.py]: Made this comment to everyone else too:
For the input try to do it in this format:
var_name : type - description (eg start : integer - id of the start point)

For the output do this format:
type - description (eg list - a list of the distances between point and line

File: code/xmlParser.py [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+        for id in range(start+1, end):
+            # do side test (return lenght of line x perpendicualr distace) so devide it by the length and voila
+            diff = abs(misc.side_test((p_start[0], p_start[2]), (p_end[0], p_end[2]), (self.vts[id,0], self.vts[id,2]))) / line_length
+            offsets.append(diff)
+
+        return np.array(offsets)
+
+    def douglas_Peucker(self, threshold):
+        # initalize simple path first first and last point
+        path_simple = [0, len(self.vts)-1]
```
Comment (Constantijn-Dinklo) [inline code/xmlParser.py]: Needs initial comment statement.

File: code/xmlParser.py [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+    def write_xml(self, filename, validate):
+        """
+        Explination:
+            1. create a element tree and set general information
+            2. insert points in tree
+            3. set point types (source, receiver etc)
+            4. write to file
+        ---------------
+        Input: 
+            filename of the output file (including the path)
```
Comment (Constantijn-Dinklo) [inline code/xmlParser.py:120]: Missing an input.

File: code/xmlParser.py [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+            id_max = np.argmax(offsets)
+
+            # Check if the offset is above the treshold, if so, add the point to the list
+            if(offsets[id_max] > threshold):
+                # Make sure to get the right id, id_max starts at 0, but 0 is already 1 further than the start point.
+                path_simple.insert(i+1, id_max + start + 1)
+            else:
+                i += 1
+        
+        self.vts_simple = np.array([self.vts[id] for id in path_simple])
```
Comment (Constantijn-Dinklo) [inline code/xmlParser.py]: Any self variable should be defined in the __init__ function.

File: code/xmlParser.py [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+
+    #xml_section = XmlParser(vertices, Materials, extension)
+    xml_section = XmlParser(vertices, Materials)
+    
+    xml_section.set_coordinates_local()
+
+    # For direct path only:
+    xml_section.unfold_straight_path()
+
+    xml_section.douglas_Peucker(0.3)
```
Comment (Constantijn-Dinklo) [inline code/xmlParser.py]: As far as I can tell, these calls don't do anything right now. This is fine! Just making sure that that is okay and you're not missing some part of the steps.

**Review by Constantijn-Dinklo — COMMENTED** [review]
Please look at the comments I made in some files.
Also, please remove any files that are not part of this commit. Only files that have anything to do with the xmlParsing process should be commited through this pull request.

File: code/misc.py (hunk 2 above)
Comment (LaurensVanRijssel) [reply code/misc.py:24]: Correct, it is not used. But it can be an input.
In put x, y, z points into the function. This does not change anything. But I added this to show that bith 2D and 3D vectors are supported

File: code/groundTin.py (hunk 3 above)
Comment (LaurensVanRijssel) [reply code/groundTin.py]: It is already updated in crossSection, I removed it.

File: code/xmlParser.py (hunk 4 above)
Comment (LaurensVanRijssel) [reply code/xmlParser.py]: ah, yeah, that sounds clearer

File: code/xmlParser.py (hunk 5 above)
Comment (LaurensVanRijssel) [reply code/xmlParser.py]: adjusted it, now correct

File: code/xmlParser.py (hunk 11 above)
Comment (LaurensVanRijssel) [reply code/xmlParser.py]: yeah, I noticed. I have been playing around with it, but didn't check at the end.

File: code/xmlParser.py (hunk 6 above)
Comment (LaurensVanRijssel) [reply code/xmlParser.py]: Nonetheless whether the path is adjusted to d, y, z, I want the write_xml to work the same. So wanted te stick to self.vts. 
In write_xml for each vertex all three values are written, where [2] is the z value. so, z has to stay at the third place. That is why I decided to do it like this. However, let me check the documentation on whether this is actually usefull to do other that it is easier humanly readable

Comment (LaurensVanRijssel) [reply code/xmlParser.py]: I checked the documentation. this will not make the code faster for xml. So I deleted the function
